File: server/src/main/java/com/oreflow/auroreflow/services/LightbulbService.java
/**
 * Copyright 2017 Tim Malmström
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.oreflow.auroreflow.services;

import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableList;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.oreflow.auroreflow.proto.AuroreflowProto;
import com.oreflow.auroreflow.proto.AuroreflowProto.LightbulbRequest;
import com.oreflow.auroreflow.proto.AuroreflowProto.Lightbulb;

import java.io.*;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Level;
import java.util.logging.Logger;


/**
 * Service to manage Lightbulbs and their current states
 */
@Singleton
public class LightbulbService {
  private static final Logger logger = Logger.getLogger(LightbulbService.class.getName());

  private final WebsocketService websocketService;
  private final Map<Long, Lightbulb> lightbulbs;

  @Inject
  LightbulbService(WebsocketService websocketService) {
    this.websocketService = websocketService;
    lightbulbs = new ConcurrentHashMap<>();
  }

  /**
   * Adds a new {@link Lightbulb}
   */
  void putLightbulb(Lightbulb lightbulb) throws IOException {
    updateLightbulb(lightbulb);
  }

  /**
   * Gets a {@link Lightbulb} by Id, or throws exception if it does not exist
   */
  public Lightbulb getLightbulb(long id) {
    if (lightbulbs.containsKey(id)) {
      return lightbulbs.get(id);
    }
    throw new IllegalArgumentException("Lightbulb ID does not exist " + id);
  }

  /**
   * Updates the stored {@link Lightbulb} with information in given {@link LightbulbRequest}
   */
  void updateLightbulbWith(Lightbulb lightbulb, LightbulbRequest lightbulbRequest) {
    Lightbulb.Builder updatedLightbulb = lightbulb.toBuilder()
        .setLastChangeMillis(lightbulbRequest.getRequestTime())
        .setIsActive(true);
    switch (lightbulbRequest.getRequestTypeCase()) {
      case HSV_REQUEST:
        updateLightbulb(
            updatedLightbulb
                .setHue(lightbulbRequest.getHsvRequest().getHue())
                .setSat(lightbulbRequest.getHsvRequest().getSat())
                .setBright(lightbulbRequest.getHsvRequest().getBrightness())
                .setColorMode(Lightbulb.ColorMode.COLOR_MODE)
                .setPower(AuroreflowProto.Power.ON)
                .build());
        break;
      case CT_REQUEST:
        updateLightbulb(
            updatedLightbulb
                .setCt(lightbulbRequest.getCtRequest().getCt())
                .setBright(lightbulbRequest.getCtRequest().getBrightness())
                .setColorMode(Lightbulb.ColorMode.COLOR_TEMPERATURE_MODE)
                .setPower(AuroreflowProto.Power.ON)
                .build());
        break;
      case POWER_REQUEST:
        updateLightbulb(
            updatedLightbulb
                .setPower(lightbulbRequest.getPowerRequest().getPower())
                .build());
        break;
      case NAME_REQUEST:
        updateLightbulb(
            updatedLightbulb
                .setName(lightbulbRequest.getNameRequest().getName())
                .build());
        break;
      case REQUESTTYPE_NOT_SET:
        updateLightbulb(lightbulb.toBuilder().setIsActive(true).build());
        break;
    }
  }

  /**
   * Updates stored {@link Lightbulb} and triggers weboscket broadcast about the update
   */
  void updateLightbulb(Lightbulb lightbulb) {
    if(!lightbulbs.containsKey(lightbulb.getId())
        || !lightbulbs.get(lightbulb.getId()).equals(lightbulb)) {
      logger.log(Level.INFO, String.format("UPDATE ID: %d WITH STATUS: %b", lightbulb.getId(), lightbulb.getIsActive()));
      lightbulbs.put(lightbulb.getId(), lightbulb);
      websocketService.broadcastLightbulbUpdate(lightbulb);
    }
  }

  /**
   * Checks if the service currently has a lightbulb of provided ID
   */
  boolean hasLightbulb(long lightbulbId) {
    return lightbulbs.containsKey(lightbulbId);
  }

  /**
   * Gets a list of all {@link Lightbulb}s
   */
  public ImmutableCollection<Lightbulb> getAllLightbulbs() {
    return ImmutableList.copyOf(lightbulbs.values());
  }
}

File: web/src/app/service/lightbulb.service.ts
/**
 * Copyright 2017 Tim Malmström
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';

import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/map';

import { Lightbulb } from '../model/lightbulb';
import {
    CtRequest,
    HsvRequest,
    NameRequest,
    PowerRequest
} from "../model/lightbulbrequest";
import { Subject } from 'rxjs/Subject';

@Injectable()
export class LightbulbService {
    private PUT_HEADERS = new Headers({ 'Content-Type': 'application/json' });
    lightbulbSubject: Subject<Lightbulb[]> = new Subject();
    lightbulbs: Lightbulb[];
    private websocket: WebSocket;

    constructor(private http: HttpClient) {
        this.connectWebsocket();
        this.getLightbulbs(true);
    }

    connectWebsocket() {
        if(this.websocket) {
            this.websocket.close();
        } 
        this.websocket = new WebSocket('ws://' + window.location.host + '/ws');
        this.websocket.onmessage = (message) => this.bulbUpdateHandler(message);
    }

    bulbUpdateHandler(message: MessageEvent) {
        const lightbulb: Lightbulb = JSON.parse(message.data);
        const existingBulb = this.lightbulbs.find(bulb => bulb.Id == lightbulb.Id);
        if(existingBulb) {
            if (JSON.stringify(existingBulb) == JSON.stringify(lightbulb)) {
                return;
            } 
            existingBulb.Power = lightbulb.Power;
            existingBulb.Bright = lightbulb.Bright;
            existingBulb.Ct = lightbulb.Ct;
            existingBulb.Hue = lightbulb.Hue;
            existingBulb.Sat = lightbulb.Sat;
            existingBulb.ColorMode = lightbulb.ColorMode;
            existingBulb.Name = lightbulb.Name;
            existingBulb.IsActive = lightbulb.IsActive;
        } else {
            this.lightbulbs.push(lightbulb);
        }
        this.lightbulbs = this.lightbulbs.sort((bulbA, bulbB) => bulbA.Name.localeCompare(bulbB.Name));
        this.lightbulbSubject.next(this.lightbulbs);
    }

    getLightbulbs(forceNext: boolean) {
        this.http.get<Lightbulb[]>('lightbulb/list').subscribe(data => {
            if (forceNext || !this.lightbulbs || data.length != this.lightbulbs.length) {
                this.lightbulbs = data.sort((bulbA, bulbB) => bulbA.Name.localeCompare(bulbB.Name));
                this.lightbulbSubject.next(this.lightbulbs);
            }
        });
        return this.lightbulbSubject.asObservable();
    }

    /**
     * Sends a Hsv update request using the HSV values provided in the Lightbulb object
     */
    sendHsvUpdate(lightbulb: Lightbulb): void {
        lightbulb.ColorMode = 2;
        const request: HsvRequest  = {
            Id: lightbulb.Id,
            Hue: lightbulb.Hue,
            Sat: lightbulb.Sat,
            Bright: lightbulb.Bright
        }
        this.http.put('lightbulb/update/hsv', JSON.stringify(request)).subscribe();
    }

    /**
     * Sends a Ct update request using the Ct values provided in the Lightbulb object
     */
    sendCtUpdate(lightbulb: Lightbulb): void {
        lightbulb.ColorMode = 1;
        const request: CtRequest  = {
            Id: lightbulb.Id,
            Ct: lightbulb.Ct,
            Bright: lightbulb.Bright
        }
        this.http.put('lightbulb/update/ct', JSON.stringify(request)).subscribe();
    }

    /**
     * Sends a Power update request using the power value provided in the Lightbulb object
     */
    sendPowerUpdate(lightbulb: Lightbulb): void {
        const request: PowerRequest = {
            Id: lightbulb.Id,
            Power: lightbulb.Power
        };
        this.http.put('lightbulb/update/power', JSON.stringify(request)).subscribe();
    }

    /**
     * Sends a Name update request using the name value provided in the Lightbulb object
     */
    sendNameUpdate(lightbulb: Lightbulb): void {
    }

    /**
     * Sends a power off all bulbs request
     */
    sendPowerOffAll(): void {
        this.http.get('poweroff').subscribe();
    }
}